src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2362,148 +2392,117 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
                 return;
             }
 
-            // For certain variables, we always know their type, well at least we can assume we know.
-            if (astVariablePath.IsUnqualified)
+            // Process well known variables like: $this and $true
+            if (astVariablePath.IsUnqualified
+                && astVariablePath.UserPath.EqualsOrdinalIgnoreCase(SpecialVariables.This)
+                && (_context.CurrentTypeDefinitionAst is not null || _context.CurrentThisType is not null))
             {
-                var isThis = astVariablePath.UserPath.EqualsOrdinalIgnoreCase(SpecialVariables.This);
-                if (!isThis || (_context.CurrentTypeDefinitionAst == null && _context.CurrentThisType == null))
+                // $this is special in script properties and in PowerShell classes
+                PSTypeName typeName = _context.CurrentThisType ?? new PSTypeName(_context.CurrentTypeDefinitionAst);
+                inferredTypes.Add(typeName);
+                return;
+            }
+            else
+            {
+                if (SpecialVariables.AllScopeVariables.TryGetValue(astVariablePath.UnqualifiedPath, out Type knownType))
                 {
-                    if (SpecialVariables.AllScopeVariables.TryGetValue(astVariablePath.UserPath, out Type knownType))
+                    if (knownType == typeof(object))
                     {
-                        if (knownType == typeof(object))
+                        if (_context.TryGetRepresentativeTypeNameFromExpressionSafeEval(variableExpressionAst, out var psType))
                         {
-                            if (_context.TryGetRepresentativeTypeNameFromExpressionSafeEval(variableExpressionAst, out var psType))
-                            {
-                                inferredTypes.Add(psType);
-                            }
+                            inferredTypes.Add(psType);
                         }
-                        else
-                        {
-                            inferredTypes.Add(new PSTypeName(knownType));
-                        }
-
-                        return;
                     }
-
-                    for (int i = 0; i < SpecialVariables.AutomaticVariables.Length; i++)
+                    else
                     {
-                        if (!astVariablePath.UserPath.EqualsOrdinalIgnoreCase(SpecialVariables.AutomaticVariables[i]))
-                        {
-                            continue;
-                        }
-
-                        var type = SpecialVariables.AutomaticVariableTypes[i];
-                        if (type != typeof(object))
-                        {
-                            inferredTypes.Add(new PSTypeName(type));
-                        }
-
-                        break;
```

Comment:
Why do we remove the `break`?

Reply:
If we get here the variable we are analyzing is one of the automatic variables. Most of them have a type defined, and it's added in the previous if statement so there's no more work to be done for these. The only ones that don't are `$_`, `$this` and `$input`. The first 2 are handled earlier in the variable type inference so they wouldn't even get to this point. That leaves `$input` which can't really be inferred because you'd use it inside a function definition, but the value depends on the pipeline input when the function is actually called.

I can add the return inside the if statement and revert this statement to break but we still wouldn't be able to infer the value of `$input`.

Reply:
Not as far as I know.